### PR TITLE
iox-#1794 Move `uniqueMergeSortedContainers` to `WaitSet`

### LINKED ISSUE #1794: Refactor `uniqueMergeSortedContainers`
## Brief feature description

The templated function `uniqueMergeSortedContainers` in `algorithm.hpp` is only used in the `WaitSet`. While it's in general good to write generic and re-usable code, in this case it can lead to an issue when using it together with the static `iox::vector<T, N>`.

## Detailed information

Here

https://github.com/eclipse-iceoryx/iceoryx/blob/47feaef132ace0449bc84866a2688cddff1dc263/iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/algorithm.inl#L101

and in a few other places, the return value is discarded and silently ignored if a `iox::vector` is used and the returned vector is already full. As a result the user might not be aware that elements are missing and the `iox::vector` was not large enough to store the result. This behaviour might be similar with other containers, which have an `emplace_back` method, but could also differ.

* [ ] Move `uniqueMergeSortedContainers` to the `WaitSet` and de-templatize it

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

* Move `uniqueMergeSortedContainers` to the `WaitSet` and de-templatize it

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1794

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/internal/popo/wait_set.inl [hunk 1]
```diff
@@ -317,10 +317,77 @@ inline void WaitSet<Capacity>::acquireNotifications(const WaitFunction& wait) no
...
+                {
+                    IOX_LOG(WARN) << "Overflow when merging new notifications with already active notification! "
+                                     "Consider increasing the CMake switch 'IOX_INTERNAL_MAX_NUMBER_OF_NOTIFIERS'";
+                }
+            });
     }
 }
 
+template <uint64_t Capacity>
+inline cxx::expected<ConditionListener::NotificationVector_t, NotificationVectorMergeError>
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/wait_set.inl:333]: Please do not adjust the API here. This is an internal algorithm of the waitset which is guaranteed to work due to design.
Another thing is, they callee can ensure that this works correctly by providing two vectors where the sum of all elements does not exceed the capacity.

Furthermore, we can also state in the internal doc, merges two vectors until capacity is reached so that user knows that it should be ensured by other means that this does not truncate.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/wait_set.inl [hunk 2]
```diff
@@ -317,10 +317,77 @@ inline void WaitSet<Capacity>::acquireNotifications(const WaitFunction& wait) no
     }
     else if (!notificationVector.empty())
     {
-        m_activeNotifications = algorithm::uniqueMergeSortedContainers(notificationVector, m_activeNotifications);
+        uniqueMergeSortedNotificationVector(notificationVector, m_activeNotifications)
+            .and_then([&](auto& mergedNotifications) { m_activeNotifications = mergedNotifications; })
+            .or_else([](auto& error) {
+                if (error == NotificationVectorMergeError::MERGED_VECTOR_OVERFLOW)
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/wait_set.inl:323]: Please remove this, this is unreachable and untestable code. Please stick to the original API of the merge function - it is internal API and does not require an expected.
The contract is that it merges two containers until there is no more space left. Then it truncates

**Review by elfenpiff — CHANGES_REQUESTED** [review]

File: iceoryx_posh/include/iceoryx_posh/internal/popo/wait_set.inl [hunk 3]
```diff
@@ -317,10 +317,77 @@ inline void WaitSet<Capacity>::acquireNotifications(const WaitFunction& wait) no
...
+                {
+                    IOX_LOG(WARN) << "Overflow when merging new notifications with already active notification! "
+                                     "Consider increasing the CMake switch 'IOX_INTERNAL_MAX_NUMBER_OF_NOTIFIERS'";
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/wait_set.inl:326]: @FerdinandSpitzschnueffler I might be wrong, but IIRC the [comment](https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_posh/cmake/IceoryxPoshDeployment.cmake#L65) hails from the fact that we have the ugly `do_not_touch_me` structs in the C binding. @elfenpiff Is that assumption correct?


> Merging two of them (same capacity ensured by design) can in "the worst case" therefore only lead to a vector completely filled with consecutive numbers

Ok, that makes sense. I reverted the changes.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/wait_set.inl (hunk 1 above)
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/wait_set.inl:333]: @elfenpiff Sure, got it. Reason is stated [here](https://github.com/eclipse-iceoryx/iceoryx/pull/1800#discussion_r1028196890). As this method is now private API of the `WaitSet` I don't think we'll need any doxygen or docs about it.

File: iceoryx_hoofs/test/moduletests/test_cxx_algorithm.cpp [hunk 4]
```diff
@@ -113,220 +113,4 @@ TEST_F(algorithm_test, DoesContainValue_ValueListOfMultipleValuesDoesContainValu
     ::testing::Test::RecordProperty("TEST_ID", "64c87a80-e83b-4e70-8f76-476f24804f19");
     EXPECT_TRUE(doesContainValue(7353, 42, 73, 7353));
 }
-
-TEST_F(algorithm_test, MergeTwoDisjunctNonEmptySortedContainers)
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/test/moduletests/test_cxx_algorithm.cpp:117]: @elfenpiff Fair point. As `uniqueMergeSortedContainers` is a private member method now, I'm wondering if the effort makes sense writing a mock to do a white box testing. Typically we don't and I'm sure this method is already indirectly covered by some other tests :thinking:

The coverage is about the same, difference is only due to the timing test issue:

image


Are you sure about that?

File: iceoryx_posh/include/iceoryx_posh/internal/popo/wait_set.inl [hunk 5]
```diff
@@ -317,10 +317,56 @@ inline void WaitSet<Capacity>::acquireNotifications(const WaitFunction& wait) no
...
+        m_activeNotifications = uniqueMergeSortedNotificationVector(notificationVector, m_activeNotifications);
     }
 }
 
+template <uint64_t Capacity>
+inline ConditionListener::NotificationVector_t
+WaitSet<Capacity>::uniqueMergeSortedNotificationVector(const ConditionListener::NotificationVector_t& v1,
+                                                       const ConditionListener::NotificationVector_t& v2) noexcept
+{
+    ConditionListener::NotificationVector_t mergedContainer;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/wait_set.inl:329]: Could be renamed to `mergedVector` for consistency.

File: iceoryx_hoofs/test/moduletests/test_cxx_algorithm.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/test/moduletests/test_cxx_algorithm.cpp:117]: @elfenpiff @mossmaurice lets try to kickstart this PR :)

Let me see if I get your points. On the one hand this is an implementation detail and we also don't do whitebox testing for other classes. On the other hand this is quite a critical part and it is easy to break it when there is a refactoring and the `WaitSet` tests do not check for the edge cases of this particular algorithm.

For me this leads to two options
- make it a free function in the `details`/`internal` namespace and move these tests to the `WaitSet` tests
- check if the `WaitSet` needs more tests to also cover the edge cases of this algorithm

[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/test/moduletests/test_cxx_algorithm.cpp:117]: @elBoberido 

I had a chat with @elfenpiff and we concluded that  Option 1)

> make it a free function in the details/internal namespace and move these tests to the WaitSet tests

is probably the better solution. I'll update the PR in the next few days.

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — DISMISSED** [review]

File: iceoryx_posh/test/moduletests/test_popo_waitset.cpp [hunk 6]
```diff
@@ -1536,4 +1537,229 @@ TEST_F(WaitSet_test, TimedWaitUnblocksAfterMarkForDestructionCall)
     t.join();
 }
 
+TEST_F(WaitSet_test, MergeTwoDisjunctNonEmptySortedNotificationVectors)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "4f39641f-de8a-434a-8a50-cd2b66b476da");
+    using type = ConditionListener::NotificationVector_t::value_type;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/test/moduletests/test_popo_waitset.cpp:1543]: Can't this be defined in `WaitSet_test`?

**Review by FerdinandSpitzschnueffler — DISMISSED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]
